source/extensions/common/async_files/async_file_context_basic.cc

```diff
@@ -0,0 +1,354 @@
+#include "source/extensions/common/async_files/async_file_context_basic.h"
+
+#include <fcntl.h>
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/extensions/common/async_files/async_file_action.h"
+#include "source/extensions/common/async_files/async_file_context_impl.h"
+#include "source/extensions/common/async_files/status_after_file_error.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace AsyncFiles {
+
+namespace {
+class AsyncFileActionImpl : public AsyncFileAction {
+protected:
+  int& fileDescriptor(AsyncFileContextImpl* context) {
+    return static_cast<AsyncFileContextBasic*>(context)->file_descriptor_;
+  }
+};
+
+class ActionCreateAnonymousFile : public AsyncFileActionImpl {
+public:
+  ActionCreateAnonymousFile(std::string path, std::function<void(absl::Status)> on_complete)
+      : on_complete_(on_complete), path_(path) {}
+
+  void execute(AsyncFileContextImpl* context) override {
+    if (fileDescriptor(context) != -1) {
+      result_ = absl::AlreadyExistsError(
+          "AsyncFileContext::createAnonymousFile: a file was already opened in this context");
+      return;
+    }
+    bool was_successful_first_call = false;
+    static bool supports_O_TMPFILE;
+    static std::once_flag once_flag;
+    std::call_once(once_flag, [this, &context, &was_successful_first_call]() {
+      fileDescriptor(context) = open(path_.c_str(), O_TMPFILE | O_RDWR, S_IRUSR | S_IWUSR);
+      was_successful_first_call = supports_O_TMPFILE = (fileDescriptor(context) != -1);
+    });
+    if (supports_O_TMPFILE) {
+      if (was_successful_first_call) {
+        // This was the thread doing the very first open(O_TMPFILE), and it worked, so no need to do
+        // anything else.
+        return;
+      }
+      // This was any other thread, but O_TMPFILE proved it worked, so we can do it again.
+      fileDescriptor(context) = open(path_.c_str(), O_TMPFILE | O_RDWR, S_IRUSR | S_IWUSR);
+      if (fileDescriptor(context) == -1) {
+        result_ = statusAfterFileError();
+      }
+      return;
+    }
+    // If O_TMPFILE didn't work, fall back to creating a named file and unlinking it.
+    // Use a fixed-size buffer because we're going to be using C file functions anyway, and it saves
+    // a heap allocation.
+    char filename[4096];
+    static const char file_suffix[] = "/buffer.XXXXXX";
+    if (path_.size() + sizeof(file_suffix) > sizeof(filename)) {
+      result_ = absl::InvalidArgumentError(
+          "AsyncFileContext::createAnonymousFile: pathname too long for tmpfile");
+      return;
+    }
+    snprintf(filename, sizeof(filename), "%s%s", path_.c_str(), file_suffix);
```

Comment:
Please remove all C style string handling here and elsewhere. You can use absl::StrCat, absl::StrFormat, etc. for better safety.

Reply:
As discussed, mkstemp explicitly requires a C-style string because it modifies the input string in-place. So while it would be possible to StrCat and then copy that into a C-style string, that operation might as well be done in one.